### ISSUE TITLE
[WIP] failed codegen

### ISSUE THREAD

**Issue by jjsjann123** [csarofeen/pytorch#962]
## 🐛 Bug

Another codegen failure in JOC bert with AMP + nvfuser

## To Reproduce

I'm logging the graph for further investigation. I should be able to extract this into a python test.

Error I'm seeing is
```
RuntimeError: !gpu_lower->caIndexMap().areMapped( (*it_i)->iter_domain(), (*it_j)->iter_domain())INTERNAL ASSERT FAILED at "../torch/csrc/jit/codegen/cuda/index_reference_replay.cpp":106, please report a bug to PyTorch. Unsupported loop structure. Two loops are mapped together.
```

```
graph(%10 : Half(1024, strides=[1], requires_grad=0, device=cuda:0),
      %12 : Half(1024, strides=[1], requires_grad=0, device=cuda:0),
      %22 : Float(32, 128, 1024, strides=[131072, 1024, 1], requires_grad=0, device=cuda:0),
      %31 : Half(32, 128, 1024, strides=[131072, 1024, 1], requires_grad=0, device=cuda:0),
      %34 : Half(1024, strides=[1], requires_grad=0, device=cuda:0)):
  %6 : float = prim::Constant[value=9.9999999999999998e-13]()
  %3 : int[] = prim::Constant[value=[1024]]()
  %18 : int = prim::Constant[value=1]() # <string>:76:118
  %27 : bool = prim::Constant[value=1]()
  %26 : float = prim::Constant[value=1.1111111111111112]()
  %25 : float = prim::Constant[value=0.90000000000000002]()
  %35 : Half(1024, strides=[1], requires_grad=0, device=cuda:0) = aten::autocast_to_fp16(%34) # <string>:30:19
  %33 : Half(32, 128, 1024, strides=[131072, 1024, 1], requires_grad=0, device=cuda:0) = prim::add_optional(%31, %35)
  %res.1 : Half(32, 128, 1024, strides=[131072, 1024, 1], requires_grad=0, device=cuda:0), %mask.8 : Bool(32, 128, 1024, strides=[131072, 1024, 1], requires_grad=0, device=cuda:0) = aten::native_dropout(%33, %25, %26, %27) # <string>:46:23
  %23 : Float(32, 128, 1024, strides=[131072, 1024, 1], requires_grad=0, device=cuda:0) = aten::autocast_to_fp32(%22) # <string>:23:19
  %21 : Float(32, 128, 1024, strides=[131072, 1024, 1], requires_grad=0, device=cuda:0) = aten::autocast_to_fp32(%res.1)
  %19 : Float(32, 128, 1024, strides=[131072, 1024, 1], requires_grad=0, device=cuda:0) = aten::add(%21, %23, %18) # /my_workspace/DeepLearningExamples/PyTorch/LanguageModeling/BERT/modeling.py:374:39
  %15 : Float(32, 128, 1024, strides=[131072, 1024, 1], requires_grad=0, device=cuda:0) = aten::autocast_to_fp32(%19) # <string>:23:19
  %13 : Float(1024, strides=[1], requires_grad=0, device=cuda:0) = aten::autocast_to_fp32(%12) # <string>:23:19
  %11 : Float(1024, strides=[1], requires_grad=0, device=cuda:0) = aten::autocast_to_fp32(%10) # <string>:23:19
  %output.1 : Float(32, 128, 1024, strides=[131072, 1024, 1], requires_grad=0, device=cuda:0), %mean.8 : Float(32, 128, 1, strides=[128, 1, 1], requires_grad=0, device=cuda:0), %rstd.8 : Float(32, 128, 1, strides=[128, 1, 1], requires_grad=0, device=cuda:0) = aten::native_layer_norm(%15, %3, %11, %13, %6) # <string>:33:33
  %1 : Half(32, 128, 1024, strides=[131072, 1024, 1], requires_grad=0, device=cuda:0) = aten::autocast_to_fp16(%output.1) # <string>:30:19
  return (%1, %output.1, %mean.8, %rstd.8, %11, %13, %15, %19, %res.1, %mask.8, %35)
```

**Comment by jjsjann123** [comment]
Got python repro

```
import torch                                                                                                               
import time                                                                                                                
from torch.nn import functional as F                                                                                       
                                                                                                                           
class MyModule(torch.nn.Module):                                                                                           
  def __init__(self):                                                                                                      
    super(MyModule, self).__init__()                                                                                       
    self.ln = torch.nn.Linear(1024, 1024)                                                                                  
    self.dropout = torch.nn.Dropout()                                                                                      
    self.layernorm = torch.nn.LayerNorm((1024,))                                                                           
                                                                                                                           
  def forward(self, x, y):                                                                                                 
    x = self.ln(x)                                                                                                         
    res1 = self.dropout(x)                                                                                                 
    comb = res1.add(y)                                                                                                     
    o = self.layernorm(comb)                                                                                               
    return o                                                                                                               
                                                                                                                           
shape = (32, 128, 1024)                                                                                                    
x = torch.rand(*shape).half().cuda().requires_grad_()                                                                      
y = torch.rand(*shape).float().cuda().requires_grad_()                                                                     
                                                                                                                           
forward = MyModule().cuda().half()                                                                                         
jit_forward = torch.jit.script(forward)                                                                                    
                                                                                                                           
for k in range(3):                                                                                                         
  with torch.jit.fuser("fuser2"):                                                                                          
    with torch.cuda.amp.autocast(enabled=True):                                                                            
      loss = jit_forward(x, y)                                                                                             
                                                                                                                           
with torch.cuda.amp.autocast(enabled=True):                                                                                
  print(jit_forward.graph_for(x, y))
```

**Comment by jjsjann123** [comment]
An even smaller repro from @kevinstephano 
```
import torch                                                                                                               
import time                                                                                                                
from torch.nn import functional as F                                                                                       
                                                                                                                           
import os                                                                                                                  
                                                                                                                           
os.environ['PYTORCH_NVFUSER_DISABLE_FALLBACK'] = '1'                                                                                                                                          
                                                                                                                           
class MyModule(torch.nn.Module):                                                                                           
  def __init__(self):                                                                                                      
    super(MyModule, self).__init__()                                                                                       
    self.layernorm = torch.nn.LayerNorm((1024,))                                                                           
                                                                                                                           
  def forward(self, x, y):                                                                                                 
    comb = x.add(y)                                                                                                        
    o = self.layernorm(comb)                                                                                               
    return o                                                                                                               
                                                                                                                           
x = torch.rand((32, 128, 1024)).half().cuda().requires_grad_()                                                             
y = torch.rand(1024).float().cuda().requires_grad_()                                                                       
                                                                                                                           
forward = MyModule().cuda().half()                                                                                         
jit_forward = torch.jit.script(forward)                                                                                    
                                                                                                                           
for k in range(3):                                                                                                         
  with torch.jit.fuser("fuser2"):                                                                                          
    with torch.cuda.amp.autocast(enabled=True):                                                                            
      loss = jit_forward(x, y)                                                                                             
                                                                                                                           
with torch.cuda.amp.autocast(enabled=True):                                                                                
  print(jit_forward.graph_for(x, y))
```

**Comment by jjsjann123** [comment]
Forgot to mention the update to autocast. This seems to be necessary to get the repro

```
jiej@yamcha:/raid/jiej/pytorch$ git diff
diff --git a/torch/csrc/jit/passes/autocast.cpp b/torch/csrc/jit/passes/autocast.cpp
index 924c33c629..9e85a64cf7 100644
--- a/torch/csrc/jit/passes/autocast.cpp
+++ b/torch/csrc/jit/passes/autocast.cpp
@@ -280,6 +280,7 @@ void handleBlock(Block* block, bool initial_state) {
         break;

       // CastPolicy::promote (promote inputs to the widest type)
+      case aten::add:
       case aten::addcdiv:
       case aten::addcmul:
       case aten::atan2:
```

**Comment by jjsjann123** [comment]
No longer repros and appears to be fixed by #1035
